### ISSUE TITLE
wrong in parameters in write_intro_doc function

### ISSUE THREAD

**Issue by Lycxder** [JaimeArboleda/nestedcvtraining#2]
When calling write_intro_doc function in api.py, there's no _ensemble_ parameter in the list which could cause the problem _write_intro_doc() missing 1 required positional argument: 'skopt_func'_ because of lack of parameters. I noticed that the _ensemble_ parameter is the same as _undersampling_majority_class_ in search space. I tried to delete the _ensemble_ from the parameter list of definition of function _write_intro_doc_ and the code blocks related to _ensemble_ in this function is also removed but it does not work. Appreciate it if you have spare time to solve the problem.

**Comment by JaimeArboleda** [comment]
Hello! If you could reproduce the error with a code sample it would be great. I have been checking and in principle it works in my case. Ensemble should exist always because it is a parameter of the find_best_binary_model function. It has a default value of False, but this should not be a problem. 

The use of this parameter is different from undersampling_majority_class when used as an option for the search space. I'll try to explain. First of all remember that we have an inner loop (with inner folds) and an outer loop (with outer folds). The inner loop has the purpose of finding the best combination of hyperparameters, and the outer loop has the purpose of estimating the error of the model. Now:
- Ensemble = True  means that, in the outer loop, once the best parameters have been found using a bayesian search over the search space, averaging over the several inner folds corresponding to this outer fold, instead of building a new model with those parameters using all data corresponding to the outer fold, an ensemble model is created (that is, a model that just averages the predicitons of all "inner" models). And if ensemble = False, then instead of an ensemble model, a new model trained on all available data for the outer fold is built, and this is the model that will be evaluated against the test data corresponding to the outer fold. 
- undersampling_majority_class is a method for dealing with imbalanced data, that consists on the following idea: if we have, let's say, 10 times more examples of label 0 than label 1, then you could train 10 models (each of them sampling 1/10 of the 0 label and using the whole minority class). And those 10 models will, in turn, work together as an ensemble. 

As you can see, those two concepts are related, but not the same. The ensemble parameter creates an ensemble of models, and the other is an algorithm for dealing with imbalanced data by creating an ensemble of models trained on balanced samples of the whole dataset. In a way, the second implies the first, but if the second is used, the first is ignored. 

I don't know if that is clear for you or not... Anyway, if you have encountered a bug and the code does not run please provide me with an example (code and data if possible) and I'll try to fix it.

**Comment by Lycxder** [comment]
Thanks a lot! I understand what you said and thank you very much for your kind and professional reply!
I just used the code you provided as [example_binary.py](https://github.com/JaimeArboleda/nestedcvtraining/blob/master/example_binary.py). 
This is my code after little modification.
`from nestedcvtraining.api import find_best_binary_model, MidasIdentity, OptionedPostProcessTransformer
y = dataset_prueba["CLASS"].to_numpy()
X = dataset_prueba[[c for c in dataset_prueba.columns if c != "CLASS"]].to_numpy()
dict_models = {
    "xgboost": {
        "model": XGBClassifier(),
        "pipeline_post_process": Pipeline(
            [
                (
                    "post_process",
                    OptionedPostProcessTransformer(dict_pipelines_post_process),
                ),
                ("resample", SMOTE()),
            ]
        ),
        "search_space": [
            Categorical([True, False], name="undersampling_majority_class"),
            Integer(5, 6, name="max_k_undersampling"),
            Categorical(["minority", "all"], name="resample__sampling_strategy"),
            Categorical(
                ["option_1", "option_2", "option_3"], name="post_process__option"
            ),
            Integer(5, 15, name="model__max_depth"),
            Real(0.05, 0.31, prior="log-uniform", name="model__learning_rate"),
            Integer(1, 10, name="model__min_child_weight"),
            Real(0.8, 1, prior="log-uniform", name="model__subsample"),
            Real(0.13, 0.8, prior="log-uniform", name="model__colsample_bytree"),
            Real(0.1, 10, prior="log-uniform", name="model__scale_pos_weight"),
            Categorical(["binary:logistic"], name="model__objective"),
        ],
    },
    "random_forest": {
        "model": RandomForestClassifier(),
        "pipeline_post_process": None,
        "search_space": [
            Categorical([True, False], name="undersampling_majority_class"),
            Integer(0, 1, name="model__bootstrap"),
            Integer(10, 100, name="model__n_estimators"),
            Integer(2, 10, name="model__max_depth"),
            Integer(5, 20, name="model__min_samples_split"),
            Integer(1, 4, name="model__min_samples_leaf"),
            Categorical(["auto", "sqrt"], name="model__max_features"),
            Categorical(
                ["balanced", "balanced_subsample"], name="model__class_weight"
            ),
        ],
    },
}

best_model, report_doc, report_dfs = find_best_binary_model(
    X=X,
    y=y,
    model_search_spaces=dict_models,
    verbose=True,
    k_inner_fold=10,
    k_outer_fold=9,
    skip_inner_folds=[0, 2, 4, 6, 8, 9],
    skip_outer_folds=[0, 2, 3, 4, 6, 8],
    n_initial_points=10,
    n_calls=10,
    build_final_model=False,
    loss_metric="average_precision",
    peeking_metrics=[
        "roc_auc",
        "neg_log_loss",
        "average_precision",
        "neg_brier_score",
    ],
    skopt_func=gbrt_minimize
)
report_doc.save("report_dataset.docx")
for model in report_dfs.keys():
    report_dfs[model].to_csv('report_dfs ' + model + '.csv', sep=';', index=False)`
My example data is attached here.
[dataset_prueba.csv](https://github.com/JaimeArboleda/nestedcvtraining/files/8493704/dataset_prueba.csv)

I would be grateful if you could help solve this problem.

**Comment by JaimeArboleda** [comment]
Thanks a lot, I am trying to run this example and I am stuck in another issues... To be fair with you, I am afraid this project needs a huge reformatting, so it will take some time. I don't know if you can afford waiting or not... But I need to be honest with you in this respect. I would like to fix this error (and others) and add some features that are lacking, but I don't have the time for that right now. I am very sorry for that... Probably I will have to wait until summer, when I have more time to invest on this.

**Comment by Lycxder** [comment]
I understand and thanks a lot for checking my code and data. After fixing the bugs if you need someone to test, please contact me by leaving the message in this issue and I would make efforts to help.

**Comment by JaimeArboleda** [comment]
Hello! I am sorry for the delay, but now I am fixing this library. In fact, I am refactoring almost every part to make it simpler and to make it work, which is the most important thing :)
It will also work on multiclass settings. 
I hope next week it will be done.

**Comment by JaimeArboleda** [comment]
I have removed the part of reporting in a doc (I find it very clumsy and I think a plain dataframe collecting all results is enough). This allowed to simplify the code a little bit, albeit I think the package needs more refactoring. Anyway, it should be functional now. If you install the new version, please tell me if you find problems or not. Thanks!

**Comment by JaimeArboleda** [comment]
I close it because in the newer versions this issue is solved.
